lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -1564,16 +1567,32 @@ checkpointFromEntity cp utxo =
         bh
         ) = cp
     header = (W.BlockHeader slot (Quantity bh) headerHash parentHeaderHash)
-    utxo' = W.UTxO . Map.fromList $
-        [ (W.TxIn input ix, W.TxOut addr (mkTokenBundle coin tokens))
-        | (UTxO _ _ (TxId input) ix addr coin, tokens) <- utxo
+    utxo = W.UTxO $ Map.merge
+        (Map.mapMissing (const mkFromCoin)) -- No assets, only coins
+        (Map.dropMissing) -- Only assets, impossible.
+        (Map.zipWithMatched (const mkFromBoth)) -- Both assets and coins
+        (Map.fromList
+            [ (W.TxIn input ix, (addr, coin))
+            | (UTxO _ _ (TxId input) ix addr coin) <- coins
+            ])
+        (Map.fromListWith TokenBundle.add
```

Comment:
Why `Map.fromListWith TokenBundle.add`, but `Map.fromList` for ada?

Reply:
Ah —  DB `UTxO` is specific to one asset, so there can be multiple ones per input, which needs to be merged.

Reply:
Precisely :+1: